### PR TITLE
Remove mutable data function in id(UniqueID and its subclass)

### PR DESCRIPTION
**Pull request by ashione**
cache hash code by member var

## What do these changes do?
* remove mutable data function in id.c/id.h
* add private constructor UniqueID(const std::string &binaray)
* remove subclass related private function

## Related issue number

https://github.com/ray-project/ray/issues/4695
## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: src/ray/id.cc [hunk 1]
```diff
@@ -24,18 +24,26 @@ std::mt19937 RandomlySeededMersenneTwister() {
...
 
 UniqueID::UniqueID(const plasma::UniqueID &from) {
   std::memcpy(&id_, from.data(), kUniqueIDSize);
+  hash_ = MurmurHash64A(&id_[0], kUniqueIDSize, 0);
 }
 
 UniqueID UniqueID::from_random() {
-  UniqueID id;
-  uint8_t *data = id.mutable_data();
+  std::string data;
```
Comment (guoyuhong) [inline src/ray/id.cc]: Will `std::string data(kUniqueIDSize, 0);` and `data[i] = ` has better performance than `data.push_back`?

File: src/ray/common/common_protocol.h [hunk 2]
```diff
@@ -104,15 +104,13 @@ string_vec_to_flatbuf(flatbuffers::FlatBufferBuilder &fbb,
 template <typename ID>
 flatbuffers::Offset<flatbuffers::String> to_flatbuf(flatbuffers::FlatBufferBuilder &fbb,
                                                     ID id) {
-  return fbb.CreateString(reinterpret_cast<const char *>(id.data()), sizeof(ID));
+  return fbb.CreateString(reinterpret_cast<const char *>(id.data()), kUniqueIDSize);
```
Comment (jovany-wang) [inline src/ray/common/common_protocol.h]: I'd like to add `kUniqueIDSize` as a static member of `ID` class, therefore, we can use it like the following code which looks more nature.
```C++
return fbb.CreateString(reinterpret_cast<const char *>(id.data()), id.kUniqueIDSize);
``` 
or
```C++
return fbb.CreateString(reinterpret_cast<const char *>(id.data()), ID::kUniqueIDSize);
```

Comment (guoyuhong) [inline src/ray/common/common_protocol.h]: Though it is the same, `kUniqueIDSize`->`id.size()` looks better? However, we cannot use `.size()` in `from_flatbuf`, which makes it a bit strange...

File: src/ray/id.h [hunk 3]
```diff
@@ -24,33 +24,35 @@ class RAY_EXPORT UniqueID {
...
   std::string binary() const;
   std::string hex() const;
   plasma::UniqueID to_plasma_id() const;
 
+ private:
+  UniqueID(const std::string &binary);
+
  protected:
   uint8_t id_[kUniqueIDSize];
+  size_t hash_;
```
Comment (jovany-wang) [inline src/ray/id.h]: Better to call it `hash_value_` or `hash_code_`?

File: src/ray/id.h [hunk 4]
```diff
@@ -24,33 +24,35 @@ class RAY_EXPORT UniqueID {
...
+#define DEFINE_UNIQUE_ID(type)                                          \
+  class RAY_EXPORT type : public UniqueID {                             \
+   public:                                                              \
+    explicit type(const UniqueID &from) {                               \
+      std::memcpy(&id_, from.data(), kUniqueIDSize);                    \
+    }                                                                   \
+    type() : UniqueID() {}                                              \
+    static type from_random() { return type(UniqueID::from_random()); } \
+    static type from_binary(const std::string &binary) {                \
+      return type(UniqueID::from_binary(binary));                       \
```
Comment (guoyuhong) [inline src/ray/id.h]: It looks like there are two copy operations: 
1. `string`->`UniqueID`
2. `UniqueID`->`ObjectID`
3. discarding `UniqueID`

How about use a private constructor and change `UniqueID`'s constructor to protected?
`type(const std::string &binary) : UniqueID(binary) {}`

**Review by jovany-wang — COMMENTED** [review]
How about to add a doc comment for `UniqueID` to note it's a immutable class?

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc [hunk 5]
```diff
@@ -228,11 +231,12 @@ Java_org_ray_runtime_raylet_RayletClientImpl_nativeGenerateTaskId(
 
   TaskID task_id =
       ray::GenerateTaskId(driver_id.GetId(), parent_task_id.GetId(), parent_task_counter);
-  jbyteArray result = env->NewByteArray(sizeof(TaskID));
+  jbyteArray result = env->NewByteArray(kUniqueIDSize);
```
Comment (guoyuhong) [inline src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: use `kUniqueIDSize`->`task_id.size()` to keep constant style as following code.

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc [hunk 6]
```diff
@@ -228,11 +231,12 @@ Java_org_ray_runtime_raylet_RayletClientImpl_nativeGenerateTaskId(
...
   if (nullptr == result) {
     return nullptr;
   }
-  env->SetByteArrayRegion(result, 0, sizeof(TaskID), reinterpret_cast<jbyte *>(&task_id));
+  env->SetByteArrayRegion(result, 0, kUniqueIDSize,
```
Comment (guoyuhong) [inline src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: The same `kUniqueIDSize`->`task_id.size()`.

**Review by guoyuhong — CHANGES_REQUESTED** [review]
It looks good! I left a few comments.

**Comment by robertnishihara** [comment]
Thanks @ashione! So based on your benchmark in https://github.com/ray-project/ray/issues/4695, this gives something like a 25% speedup for storing tons of IDs in hash tables?

File: src/ray/common/common_protocol.h (hunk 2 above)
Comment (ashione) [reply src/ray/common/common_protocol.h]: we can use id.size() in this place.

**Comment by ashione** [comment]
> Thanks @ashione! So based on your benchmark in #4695, this gives something like a 25% speedup for storing tons of IDs in hash tables?

@robertnishihara It's quite helpful if our code like this 
```cpp
std::unordered_map<UniqueID, Obj> obj_map;
....
while(loop_condition){
for(auto &id in id_vect) {
 auto &obj = obj_map[id];
...
do something
...
}
}
```

File: src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc (hunk 5 above)
Comment (ashione) [reply src/ray/raylet/lib/java/org_ray_runtime_raylet_RayletClientImpl.cc]: Sure, It should be modified in this pattern.

File: src/ray/id.h (hunk 4 above)
Comment (ashione) [reply src/ray/id.h]: There are failures in some UnitTest (lineage tests in cpp), if I try to do this. It's so weird or some implicit conversions here.

**Comment by ashione** [comment]
@guoyuhong  @robertnishihara @jovany-wang 
master VS this PR perf (run 7 times):

6094 5373 6252 5098 5397 5241 5233 (avg 5526)
4240 4748 4234 5563 4122 4014 3939 (avg 4408)

cutting cost 20.2%

**Comment by guoyuhong** [comment]
Look good to me now, but not sure whether the test failure is related or not. The test results in master branch seem stable.

**Comment by guoyuhong** [comment]
@ashione I take a look at the tests. Many test failure is caused by timeout. This PR may have some perf concerns. For many IDs we do not use the hash value. This PR will calculate the hash value no matter the hash value is needed or not. Maybe we need to update this PR to make it lazy calculation.

**Review by guoyuhong — APPROVED** [review]
After the lazy evaluation, all tests passed.

**Pull request merged**